src/loan/protocol_tests.rs

```diff
@@ -1,114 +1,87 @@
 #[cfg(test)]
 mod tests {
+    extern crate link_cplusplus;
+
+    use std::str::FromStr;
+    use std::sync::Arc;
     use std::time::SystemTime;
 
+    use crate::input::Input;
     use crate::loan::{oracle, Borrower0, Lender0};
     use anyhow::{Context, Result};
-    use bitcoin_hashes::Hash;
     use elements::bitcoin::util::psbt::serialize::Serialize;
     use elements::bitcoin::{Amount, Network, PrivateKey, PublicKey};
+    use elements::confidential::{Asset, AssetBlindingFactor, ValueBlindingFactor};
+    use elements::hashes::Hash;
     use elements::script::Builder;
     use elements::secp256k1_zkp::{SecretKey, SECP256K1};
     use elements::sighash::SigHashCache;
     use elements::{
         opcodes, Address, AddressParams, AssetId, OutPoint, Script, SigHashType, Transaction,
-        TxOut, Txid,
+        TxOut, TxOutSecrets, Txid,
     };
     use elements_harness::Elementsd;
     use elements_rpc::ElementsRpc;
-    use rand::{CryptoRng, RngCore, SeedableRng};
+    use rand::{thread_rng, CryptoRng, RngCore, SeedableRng};
     use rand_chacha::ChaChaRng;
     use testcontainers::clients::Cli;
+    use tokio::sync::Mutex;
 
     #[tokio::test]
     async fn borrow_and_repay() {
```

Comment:
@thomaseizinger I've rewritten this test to not use testcontainers, which makes it so much faster. What do you think? Should we do it for the rest and transition to this kind of testing?

Reply:
Your decision :)

To build more confidence with the lib, it may be good to include some negative tests as well. That way we know it is doing something!

Reply:
Having said that, fast and high-confidence unit tests for protocols are a little dream coming true. Can't believe it was so easy to do. So much time wasted in the past 😭